packages/overture-schema-system/src/overture/schema/system/constraint/string.py

```diff
@@ -0,0 +1,376 @@
+            context = info.context or {}
+            loc = context.get("loc_prefix", ()) + ("value",)
+            raise ValidationError.from_exception_data(
+                title=self.__class__.__name__,
+                line_errors=[
+                    InitErrorDetails(
+                        type="value_error",
+                        loc=loc,
+                        input=value,
+                        ctx={"error": self.error_message.format(value=value)},
+                    )
+                ],
+            )
+
+    def __get_pydantic_json_schema__(
+        self, core_schema: core_schema.CoreSchema, handler: GetJsonSchemaHandler
+    ) -> dict[str, Any]:
+        json_schema = handler(core_schema)
+        json_schema["pattern"] = self.pattern_str
+        return json_schema
+
+
+class LanguageTagConstraint(StringConstraint):
+    """This pattern recognizes BCP-47 language tags at the lexical or syntactic level.
+    It verifies that candidate tags follow the grammar described in the RFC, but not
+    that they are validly registered tag in IANA's language subtag registry.
+
+    In understanding the regular expression, remark that '(:?' indicates a non-capturing
+    group, and that all the top-level or non-nested groups represent top-level
+    components of `langtag` referenced in the syntax section of
+    https://www.rfc-editor.org/rfc/bcp/bcp47.txt. In particular, the top-level groups in
+    left-to-right order represent:
+
+    1. language
+    2. ["-" script]
+    3. ["-" region]
+    4. *("-" variant)
+    5. *("-" extension)
+    """
+
+    def __init__(self) -> None:
+        self.pattern = re.compile(
+            r"^(?:(?:[A-Za-z]{2,3}(?:-[A-Za-z]{3}){0,3}?)|(?:[A-Za-z]{4,8}))(?:-[A-Za-z]{4})?(?:-[A-Za-z]{2}|[0-9]{3})?(?:-(?:[A-Za-z0-9]{5,8}|[0-9][A-Za-z0-9]{3}))*(?:-[A-WY-Za-wy-z0-9](?:-[A-Za-z0-9]{2,8})+)*$"
+        )
+
+    def validate(self, value: str, info: ValidationInfo) -> None:
+        if not self.pattern.match(value):
+            context = info.context or {}
+            loc = context.get("loc_prefix", ()) + ("value",)
+            raise ValidationError.from_exception_data(
+                title=self.__class__.__name__,
+                line_errors=[
+                    InitErrorDetails(
+                        type="value_error",
+                        loc=loc,
+                        input=value,
+                        ctx={"error": f"Invalid IETF BCP-47 language tag: {value}"},
+                    )
+                ],
+            )
+
+    def __get_pydantic_json_schema__(
+        self, core_schema: core_schema.CoreSchema, handler: GetJsonSchemaHandler
+    ) -> dict[str, Any]:
+        json_schema = handler(core_schema)
+        json_schema["pattern"] = self.pattern.pattern
+        json_schema["description"] = "IETF BCP-47 language tag"
+        return json_schema
+
+
+class CountryCodeConstraint(StringConstraint):
+    """ISO 3166-1 alpha-2 country code constraint."""
```

Comment:
We should consider renaming this to reflect that it's an `alpha-2` code (to distinguish from the use of `alpha-3` codes elsewhere).

Reply:
Agree. I'll put this on the agenda.

Reply:
We're not. I'm thinking about prospective George users who don't follow our specific flavor of consistency.